src/xdist/dsession.py

```diff
@@ -205,6 +179,69 @@ def loop_once(self) -> None:
         if self.sched.tests_finished:
             self.triggershutdown()
 
+
+    def is_node_finishing(self, node: WorkerController):
+        """Check if a test worker is considered to be finishing.
+
+        Evaluate whether it's on its last test, or if no tests are pending.
+        """
+        pending = self.sched.node2pending.get(node)
+        return pending is not None and len(pending) < 2
+
+
+    def is_node_clear(self, node: WorkerController):
+        """Check if a test worker has no pending tests."""
+        pending = self.sched.node2pending.get(node)
+        return pending is None or len(pending) == 0
+
+
+    def are_all_nodes_finishing(self):
+        """Check if all workers are finishing (See 'is_node_finishing' above)."""
+        return all(self.is_node_finishing(node) for node in self.sched.nodes)
+
+
+    def are_all_nodes_done(self):
+        """Check if all nodes have reported to finish."""
+        return all(s == "finished" for s in self.worker_status.values())
+
+
+    def are_all_active_nodes_collected(self):
+        """Check if all nodes have reported collection to be complete."""
+        if not all(n.gateway.id in self.worker_status for n in self._active_nodes):
+            return False
+        return all(self.worker_status[n.gateway.id] == "collected" for n in self._active_nodes)
+
+
+    def reset_nodes_if_needed(self):
+        if self.are_all_nodes_finishing() and self.ready_to_run_tests and not self.sched.do_resched:
```

Comment:
With these changes it looks like `sched.do_resched` is no longer modified anywhere, and is by default `False`, should we just remove it?

Reply:
It should be reinstated, I removed it by mistake when squashing commits because I fail at everything

Reply:
They are reinstated now.